### PR TITLE
Avoid runtimeType.toString in toString overrides/debugLabels

### LINKED ISSUE #31695: Avoid/remove foo.runtimeType.toString()
`AbstractType_toString` is showing up in cpu profiles while scrolling on time zone selection screen on Fuchsia. This is a heavy operation and we should try to avoid it. `AbstractType_toString` is used when `toString()` is called on a result of `foo.runtimeType`.

Profiler points to the following places where this pattern is heavily used in Flutter:

1.` Widget.toStringShort()`:
```dart
String toStringShort() {
  return key == null ? '$runtimeType' : '$runtimeType-$key';
}
```
It is called from `ImplicitlyAnimatedWidgetState.initState()` at line
```dart
debugLabel: '${widget.toStringShort()}',
```

2. `BallisticScrollActivity` constructor at line
```dart
debugLabel: '$runtimeType',
```
3. Function `describeIdentity`:
```dart
String describeIdentity(Object object) => '${object.runtimeType}#${shortHash(object)}';
```

It is called from `Diagnosticable.toStringShort`:
```dart
String toStringShort() => describeIdentity(this);
```
which is called from `Diagnosticable.toString`,
which is called from `SingleTickerProviderStateMixin.createTicker`:
```dart
_ticker = Ticker(onTick, debugLabel: 'created by $this');
```

@alexmarkov

### PR DESCRIPTION
**Pull request by dnfield**
Refactoring: avoid calling `runtimeType.toString()` outside of asserts or error conditions.

~~I'll see if I can find a way to add a presubmit for this.~~ added something to the analyze.dart to check that we don't use `$runtimeType` in `toString` methods.

@a14n @zanderso @jonahwilliams 

Fixes https://github.com/flutter/flutter/issues/31695

### CODE REVIEW

File: dev/bots/analyze.dart [hunk 1]
```diff
@@ -541,6 +544,54 @@ Future<void> verifyInternationalizations() async {
   }
 }
 
+Future<void> verifyNoRuntimeTypeInToString(String workingDirectory) async {
+  final List<File> files = _allFiles(path.join(workingDirectory, 'packages', 'flutter', 'lib'), 'dart', minimumMatches: 400)
+      .where((File file) => !file.path.endsWith(path.join('lib', 'src', 'foundation', 'object.dart')))
```
Comment (jonahwilliams) [inline dev/bots/analyze.dart]: I would hoist this into a kExcludedFiles or similar

File: packages/flutter/lib/src/widgets/basic.dart [hunk 2]
```diff
@@ -2092,11 +2092,11 @@ class SizedBox extends SingleChildRenderObjectWidget {
   String toStringShort() {
     String type;
     if (width == double.infinity && height == double.infinity) {
-      type = '$runtimeType.expand';
+      type = '${objectRuntimeType(this, 'TestWidget')}.expand';
```
Comment (a14n) [inline packages/flutter/lib/src/widgets/basic.dart]: Shouldn't it be `SizedBox`?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 3]
```diff
@@ -2092,11 +2092,11 @@ class SizedBox extends SingleChildRenderObjectWidget {
...
     } else if (width == 0.0 && height == 0.0) {
-      type = '$runtimeType.shrink';
+      type = '${objectRuntimeType(this, 'TestWidget')}.shrink';
```
Comment (a14n) [inline packages/flutter/lib/src/widgets/basic.dart]: Shouldn't it be `SizedBox`?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 4]
```diff
@@ -2092,11 +2092,11 @@ class SizedBox extends SingleChildRenderObjectWidget {
...
     } else {
-      type = '$runtimeType';
+      type = '${objectRuntimeType(this, 'TestWidget')}';
```
Comment (a14n) [inline packages/flutter/lib/src/widgets/basic.dart]: Shouldn't it be `SizedBox`?

File: packages/flutter/lib/src/widgets/framework.dart [hunk 5]
```diff
@@ -3870,7 +3870,7 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
   /// A short, textual description of this element.
   @override
   String toStringShort() {
-    return widget != null ? '${widget.toStringShort()}' : '[$runtimeType]';
+    return widget != null ? '${widget.toStringShort()}' : '[${objectRuntimeType(this, '_InactiveElements')}]';
```
Comment (a14n) [inline packages/flutter/lib/src/widgets/framework.dart]: Shouldn't it be `Element`?

File: packages/flutter/lib/src/widgets/gesture_detector.dart [hunk 6]
```diff
@@ -1104,7 +1104,7 @@ abstract class SemanticsGestureDelegate {
   void assignSemantics(RenderSemanticsGestureHandler renderObject);
 
   @override
-  String toString() => '$runtimeType()';
+  String toString() => '${objectRuntimeType(this, 'GestureRecognizerFactory')}()';
```
Comment (a14n) [inline packages/flutter/lib/src/widgets/gesture_detector.dart]: Shouldn't it be `SemanticsGestureDelegate`?

File: packages/flutter/lib/src/widgets/localizations.dart [hunk 7]
```diff
@@ -135,7 +135,7 @@ abstract class LocalizationsDelegate<T> {
   Type get type => T;
 
   @override
-  String toString() => '$runtimeType[$type]';
+  String toString() => '${objectRuntimeType(this, 'Intl')}[$type]';
```
Comment (a14n) [inline packages/flutter/lib/src/widgets/localizations.dart]: Shouldn't it be `LocalizationsDelegate`?

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 8]
```diff
@@ -842,7 +843,7 @@ class _NestedScrollCoordinator implements ScrollActivityDelegate, ScrollHoldCont
   }
 
   @override
-  String toString() => '$runtimeType(outer=$_outerController; inner=$_innerController)';
+  String toString() => '${objectRuntimeType(this, 'NestedScrollView')}(outer=$_outerController; inner=$_innerController)';
```
Comment (a14n) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Shouldn't it be `_NestedScrollCoordinator`?

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 9]
```diff
@@ -1225,7 +1226,7 @@ class _NestedOuterBallisticScrollActivity extends BallisticScrollActivity {
 
   @override
   String toString() {
-    return '$runtimeType(${metrics.minRange} .. ${metrics.maxRange}; correcting by ${metrics.correctionOffset})';
+    return '${objectRuntimeType(this, '_NestedScrollController')}(${metrics.minRange} .. ${metrics.maxRange}; correcting by ${metrics.correctionOffset})';
```
Comment (a14n) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Shouldn't it be `_NestedOuterBallisticScrollActivity`?

File: packages/flutter/lib/src/widgets/routes.dart [hunk 10]
```diff
@@ -354,10 +354,10 @@ abstract class TransitionRoute<T> extends OverlayRoute<T> {
   }
 
   /// A short description of this route useful for debugging.
-  String get debugLabel => '$runtimeType';
+  String get debugLabel => '${objectRuntimeType(this, 'OverlayRoute')}';
```
Comment (a14n) [inline packages/flutter/lib/src/widgets/routes.dart]: Shouldn't it be `TransitionRoute`?

File: packages/flutter/lib/src/widgets/routes.dart [hunk 11]
```diff
@@ -354,10 +354,10 @@ abstract class TransitionRoute<T> extends OverlayRoute<T> {
...
 
   @override
-  String toString() => '$runtimeType(animation: $_controller)';
+  String toString() => '${objectRuntimeType(this, 'OverlayRoute')}(animation: $_controller)';
```
Comment (a14n) [inline packages/flutter/lib/src/widgets/routes.dart]: Shouldn't it be `TransitionRoute`?

File: packages/flutter/lib/src/widgets/routes.dart [hunk 12]
```diff
@@ -1366,7 +1366,7 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
   }
 
   @override
-  String toString() => '$runtimeType($settings, animation: $_animation)';
+  String toString() => '${objectRuntimeType(this, 'LocalHistoryEntry')}($settings, animation: $_animation)';
```
Comment (a14n) [inline packages/flutter/lib/src/widgets/routes.dart]: Shouldn't it be `ModalRoute`?

File: packages/flutter/lib/src/widgets/scroll_activity.dart [hunk 13]
```diff
@@ -507,7 +507,7 @@ class BallisticScrollActivity extends ScrollActivity {
     TickerProvider vsync,
   ) : super(delegate) {
     _controller = AnimationController.unbounded(
-      debugLabel: kDebugMode ? '$runtimeType' : null,
+      debugLabel: kDebugMode ? '${objectRuntimeType(this, 'ScrollActivityDelegate')}' : null,
```
Comment (a14n) [inline packages/flutter/lib/src/widgets/scroll_activity.dart]: Shoundn't it be `BallisticScrollActivity`?

File: packages/flutter/lib/src/widgets/scroll_metrics.dart [hunk 14]
```diff
@@ -147,6 +147,6 @@ class FixedScrollMetrics extends ScrollMetrics {
 
   @override
   String toString() {
-    return '$runtimeType(${extentBefore.toStringAsFixed(1)}..[${extentInside.toStringAsFixed(1)}]..${extentAfter.toStringAsFixed(1)})';
+    return '${objectRuntimeType(this, 'ScrollMetrics')}(${extentBefore.toStringAsFixed(1)}..[${extentInside.toStringAsFixed(1)}]..${extentAfter.toStringAsFixed(1)})';
```
Comment (a14n) [inline packages/flutter/lib/src/widgets/scroll_metrics.dart]: Shouldn't it be `FixedScrollMetrics`?

File: dev/bots/analyze.dart (hunk 1 above)
Comment (dnfield) [reply dev/bots/analyze.dart]: Done

**Comment by dnfield** [comment]
@a14n thanks for the catches - I thought I got the ones my regex missed but apparently not.  They should all be up to date now.

File: dev/bots/analyze.dart [hunk 15]
```diff
@@ -541,6 +544,58 @@ Future<void> verifyInternationalizations() async {
...
+  final List<File> files = _allFiles(flutterLib, 'dart', minimumMatches: 400)
+      .where((File file) => !excludedFiles.contains(file.path))
+      .toList();
+  final RegExp toStringRegExp = RegExp(r'^\s+String\s+to(.+?)?String(.+?)?\(\)\s+(\{|=>)');
+  final List<String> problems = <String>[];
+  for (final File file in files) {
+    final List<String> lines = file.readAsLinesSync();
+    for (int index = 0; index < lines.length; index++) {
+      if (toStringRegExp.hasMatch(lines[index])) {
+        final int sourceLine = index;
```
Comment (a14n) [inline dev/bots/analyze.dart]: `final int sourceLine = index + 1;` as lines are usually displayed 1-indexed?

**Review by a14n — APPROVED** [review]
LGTM modulo previous comment on line number in output.

File: dev/bots/analyze.dart (hunk 15 above)
Comment (dnfield) [reply dev/bots/analyze.dart]: Done, thanks for the review!

**Pull request merged**